src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyFlowEdge.cs

```diff
@@ -27,5 +28,8 @@ public DependencyFlowEdge(DependencyFlowNode from, DependencyFlowNode to, Subscr
         public bool? PartOfCycle { get; set; }
         public bool BackEdge { get; set; }
         public bool OnLongestBuildPath { get; set; }
+
+        public bool IsTooling =>
+            Subscription.LastAppliedBuild?.Dependencies.All(d => !d.IsProduct) ?? false;
```

Comment:
You should check that this does exactly what you want. My hunch is that this actually says:

- Take the last applied build (say of repo foo)
- Look up all input dependencies to repo foo's build

On other words, I think there may be two problems with this:
- If the edge is from foo->bar, this is the dependencies of the build of foo,
- The dependencies of foo's build will be a mix of toolset and product.

Reply:
@adiaaida Do you remember exactly how we proposed to determine whether the subscription only applied to product dependencies again?

My hunch is that you may be able to do the following: Look up the target repo + branch's dependency info and intersect those product dependencies with the assets produced by the last applied build of the subscription. If that set is empty, it's toolset only.

Reply:
My understanding is that we would get the source build, filter its dependent builds by the target repo/branch listed in the subscription, and grab the IsProduct. This seems to suggest that we would mark every subscription as IsTooling if one of the dependencies is a tooling dependency. Am I misreading the code?

Reply:
I think it's the other way around - if all of the dependencies are tooling dependencies then we mark subscription as `IsTooling`, if even one of them is product then `IsTooling` would be false. This snippet of code is something that I already got from Lukas before he went on holiday so maybe I just don't understand completely what is does.

Just so I get that correct: if we have subscription `foo` -> `bar` then does `LastAppliedBuild` relate to build inside `foo` or `bar` repository? And what are actually builds dependencies? 

If I should use other way of finding tooling subscription then that's fine but I would need some tips because my knowledge of Maestro data model is pretty limited.

Reply:
> I think it's the other way around - if all of the dependencies are tooling dependencies then we mark subscription as `IsTooling`, if even one of them is product then `IsTooling` would be false. This snippet of code is something that I already got from Lukas before he went on holiday so maybe I just don't understand completely what is does.

Yes, that's correct. If any is non-tooling, it's product. Though I think a 'mixed' scenario is uncommon.

> Just so I get that correct: if we have subscription foo -> bar then does LastAppliedBuild relate to build inside foo or bar repository? And what are actually builds dependencies?

the build of foo is LastAppliedBuild. And IIRC, the dependencies are the dependencies of foo itself.  So if foo dependens on A and B, it's A and B, even if foo produced C and that's what bar depends on.

Reply:
Thanks for the explanation. Then it seems that this code won't return correct information because we want to know if `foo -> bar` is tooling subscription, not only check if all dependencies of `foo` are tools.

The solution you proposed above sounds good:
> Look up the target repo + branch's dependency info and intersect those product dependencies with the assets produced by the last applied build of the subscription. If that set is empty, it's toolset only.

I will try to implement it in similar way.

Reply:
By following approach described above the (pseudo)code for determining if edge is tooling would look like this:

```
{
public async void MarkToolingEdges(IRemote barOnlyRemote)
{
    foreach (var edge in Edges)
    {
        var sourceAssets = await barOnlyRemote.GetAssetsAsync(
            buildId: edge.Subscription.LastAppliedBuild.Id);
        var dependencies = await barOnlyRemote.GetDependenciesAsync(
            repoUri: edge.To.Repository,
            branchOrCommit: edge.To.Branch);

        var hasProductDependencies = dependencies
            .Where(dependency => dependency.Type == DependencyType.Product
                && sourceAssets.Any(asset => asset.Name == dependency.Name)
            .Any();

        edge.IsTooling = !hasProductDependencies;
    }
}
```

Such naive implementation would result in two API calls per edge which seems pretty bad. On the other hand it should be pretty cheap to calculate it on database side but to do that I would need to make some changes in API. So my initial plan is to modify `channels/{id}/graph` endpoint that we already have so that it additionally returns `IsTooling` property for edges and call this endpoint from Darc instead of building graph from scratch there as we currently do.

Currently we have 3 places in code where we construct dependency graph in very similar way so it seems that it could also be beneficial to move this code to `IRemote` implementation and call it in all these places.

@mmitche @adiaaida Could you double check if what I described above makes sense before I start to write too much code? :)